osu.Game/Skinning/BeatmapSkinProvidingContainer.cs

```diff
@@ -86,22 +100,28 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
             return base.CreateChildDependencies(parent);
         }
 
+        [Resolved]
+        private SkinManager skins { get; set; }
+
         [BackgroundDependencyLoader]
-        private void load(SkinManager skins)
+        private void load()
         {
             beatmapSkins.BindValueChanged(_ => TriggerSourceChanged());
             beatmapColours.BindValueChanged(_ => TriggerSourceChanged());
             beatmapHitsounds.BindValueChanged(_ => TriggerSourceChanged());
 
-            // If the beatmap skin looks to have skinnable resources, add the default classic skin as a fallback opportunity.
-            if (skin is LegacySkinTransformer legacySkin && legacySkin.IsProvidingLegacyResources)
-            {
-                SetSources(new[]
-                {
-                    skin,
-                    skins.DefaultClassicSkin
-                });
-            }
+            if (skin != null)
+                SetSources(getBeatmapSkinSources());
+        }
+
+        private IEnumerable<ISkin> getBeatmapSkinSources()
+        {
+            ISkin transformedSkin;
+
+            yield return transformedSkin = ruleset == null ? skin : skin.WithRulesetTransformer(ruleset, beatmap.AsNonNull());
+
+            if (transformedSkin is LegacySkinTransformer legacySkin && legacySkin.IsProvidingLegacyResources)
+                yield return ruleset == null ? skins.DefaultClassicSkin : skins.DefaultClassicSkin.WithRulesetTransformer(ruleset, beatmap.AsNonNull());
```

Comment:
I'm a little confused by these `.AsNonNull()`, how can I verify they're actually correct, because they're coming into this class as nullable.

Reply:
Yeah I'm not entirely sure about the way it's done here, was mainly thinking that if `ruleset` isn't null then `beatmap` is required as well to call the ruleset transformer extension method. Alternative would be to pass the transformer method as a delegate to the container instead, think that may be better.

Reply:
If this is going to be written like this, I think I'd prefer asserts in the method.. and also `throw` or assertion on the ctor (to enforce what you stated here).